File: packages/sdk-java/src/main/java/org/melaya/MelayaEvents.java
package org.melaya;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.java_websocket.client.WebSocketClient;
import org.java_websocket.handshake.ServerHandshake;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.util.*;
import java.util.concurrent.*;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.function.Consumer;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Platform real-time events via Socket.IO at {@code /api/v1/events}.
 *
 * <p>This client implements the Engine.IO v4 / Socket.IO v4 wire protocol
 * using an HTTP long-poll handshake followed by a WebSocket upgrade (mirroring
 * the TypeScript SDK {@code events.ts}).
 *
 * <p><b>Room semantics</b> (mirror the server):
 * <ul>
 *   <li>{@code run:<runId>} — subscribe to events for a specific pipeline run</li>
 *   <li>{@code project:<project>} — subscribe to all events in a project</li>
 *   <li>{@code hitl:user:<userId>} — HITL approval events (joined automatically
 *       by the server for the authenticated user on connect)</li>
 * </ul>
 *
 * @example
 * <pre>{@code
 * MelayaEvents events = melaya.events();
 * Runnable unsub = events.onRunUpdate("run-123", frame -> System.out.println(frame));
 * events.onHitlApproval(frame -> System.out.println("HITL: " + frame));
 * // ...
 * unsub.run();      // unsubscribe
 * events.close();   // disconnect
 * }</pre>
 */
public class MelayaEvents implements AutoCloseable {

    // Engine.IO v4 packet type characters
    private static final char EIO_OPEN    = '0';
    private static final char EIO_PING    = '2';
    private static final char EIO_PONG    = '3';
    private static final char EIO_MESSAGE = '4';

    // Socket.IO packet types
    private static final int SIO_CONNECT = 0;
    private static final int SIO_EVENT   = 2;

    private static final ObjectMapper MAPPER = HttpClient.MAPPER;
    private static final Logger LOG = Logger.getLogger(MelayaEvents.class.getName());
    /** Reconnect backoff steps (ms): 1 s, 2 s, 4 s, 8 s, cap at 30 s. */
    private static final long[] RECONNECT_BACKOFF_MS = {1_000, 2_000, 4_000, 8_000, 16_000, 30_000};
    private static final int CONNECT_RETRY_MS   = 5_000;

    /** Current reconnect attempt index, reset to 0 on successful connect. */
    private volatile int reconnectAttempt = 0;

    private final String apiKey;
    private final String baseUrl;          // e.g. https://api.melaya.org
    private final java.net.http.HttpClient httpClient;

    private volatile String sid = null;
    private volatile WebSocketClient ws = null;
    private volatile boolean closed = false;
    /** Set once the first subscription triggers the lazy connect. */
    private final AtomicBoolean connectRequested = new AtomicBoolean(false);

    private final Set<String> joinedRooms = Collections.newSetFromMap(new ConcurrentHashMap<>());
    private final ConcurrentHashMap<String, Set<Consumer<JsonNode>>> listeners = new ConcurrentHashMap<>();
    /** Per-room unsubscribe handles so leaveRun/leaveProject can drop that room's listeners. */
    private final ConcurrentHashMap<String, Set<Runnable>> roomUnsubs = new ConcurrentHashMap<>();
    private final ScheduledExecutorService scheduler = Executors.newSingleThreadScheduledExecutor(r -> {
        Thread t = new Thread(r, "melaya-events");
        t.setDaemon(true);
        return t;
    });

    /**
     * Create a MelayaEvents client.
     *
     * <p>Constructing the client does <b>not</b> open a network connection.
     * The Engine.IO handshake is initiated lazily on the first subscription
     * ({@code on*}/join) call. All internal threads are daemon threads, so an
     * idle client never keeps the JVM alive.
     *
     * @param apiKey  your {@code mk_} platform API key
     * @param baseUrl REST base URL (e.g. {@code https://api.melaya.org})
     */
    public MelayaEvents(String apiKey, String baseUrl) {
        this.apiKey = apiKey;
        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.httpClient = java.net.http.HttpClient.newBuilder()
                .build();
        // Lazy: no connection is opened until the first on*/join call.
    }

    // ── Public subscription API ───────────────────────────────────────────────

    /**
     * Subscribe to run events for a specific pipeline run.
     * Automatically joins the {@code run:<runId>} Socket.IO room.
     *
     * <p>When the payload carries a {@code runId}, events for other runs are
     * filtered out, so subscribing to several runs at once never cross-delivers.
     *
     * @param runId the pipeline run ID
     * @param cb    callback invoked on each {@code pushEvent} for this run
     * @return an unsubscribe {@link Runnable} — call it to remove the listener
     */
    public Runnable onRunUpdate(String runId, Consumer<JsonNode> cb) {
        return onRoom("run:" + runId, "pushEvent", filterById("runId", "run_id", runId, cb));
    }

    /**
     * Subscribe to init-phase progress events for a run.
     * Automatically joins the {@code run:<runId>} room. Payloads carrying a
     * {@code runId} for a different run are filtered out.
     *
     * @param runId the pipeline run ID
     * @param cb    callback invoked on each {@code pushInitPhase} event
     * @return an unsubscribe {@link Runnable}
     */
    public Runnable onInitPhase(String runId, Consumer<JsonNode> cb) {
        return onRoom("run:" + runId, "pushInitPhase", filterById("runId", "run_id", runId, cb));
    }

    /**
     * Subscribe to all events in a project (all runs, pipeline CRUD).
     * Automatically joins the {@code project:<project>} room. Payloads carrying
     * a {@code project} field for a different project are filtered out.
     *
     * @param project the project name
     * @param cb      callback invoked on each {@code pushEvent} or pipeline CRUD event
     * @return an unsubscribe {@link Runnable}
     */
    public Runnable onProjectEvent(String project, Consumer<JsonNode> cb) {
        return onRoom("project:" + project, "pushEvent", filterById("project", null, project, cb));
    }

    /**
     * Subscribe to HITL approval invalidation events.
     * The server joins authenticated sockets to the HITL room automatically on connect.
     *
     * @param cb callback invoked on each {@code pushHitlApprovals} event
     * @return an unsubscribe {@link Runnable}
     */
    public Runnable onHitlApproval(Consumer<JsonNode> cb) {
        return on("pushHitlApprovals", cb);
    }

    /** Subscribe to pipeline-created events within a project room. */
    public Runnable onPipelineCreated(String project, Consumer<JsonNode> cb) {
        return onRoom("project:" + project, "pipelineCreated", filterById("project", null, project, cb));
    }

    /** Subscribe to pipeline-updated events within a project room. */
    public Runnable onPipelineUpdated(String project, Consumer<JsonNode> cb) {
        return onRoom("project:" + project, "pipelineUpdated", filterById("project", null, project, cb));
    }

    /** Subscribe to pipeline-deleted events within a project room. */
    public Runnable onPipelineDeleted(String project, Consumer<JsonNode> cb) {
        return onRoom("project:" + project, "pipelineDeleted", filterById("project", null, project, cb));
    }

    /**
     * Leave a run room and stop receiving its events.
     *
     * @param runId the pipeline run ID
     */
    public void leaveRun(String runId) {
        leaveRoom("run:" + runId);
    }

    /**
     * Leave a project room.
     *
     * @param project the project name
     */
    public void leaveProject(String project) {
        leaveRoom("project:" + project);
    }

    /** Close the Socket.IO connection and release all listeners. */
    @Override
    public void close() {
        closed = true;
        scheduler.shutdownNow();
        WebSocketClient w = ws;
        if (w != null) {
            try { w.closeBlocking(); } catch (Exception ignored) {}
        }
        ws = null;
        listeners.clear();
        joinedRooms.clear();
        roomUnsubs.clear();
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    /** Register a raw Socket.IO event listener. Returns an unsubscribe Runnable. */
    private Runnable on(String event, Consumer<JsonNode> cb) {
        ensureConnect();
        listeners.computeIfAbsent(event, k -> Collections.newSetFromMap(new ConcurrentHashMap<>())).add(cb);
        return () -> {
            Set<Consumer<JsonNode>> set = listeners.get(event);
            if (set != null) set.remove(cb);
        };
    }

    /**
     * Join a room, register a listener, and track the unsubscribe handle per
     * room so {@link #leaveRun}/{@link #leaveProject} can remove it.
     */
    private Runnable onRoom(String room, String event, Consumer<JsonNode> cb) {
        joinRoom(room);
        Runnable unsub = on(event, cb);
        Set<Runnable> set = roomUnsubs.computeIfAbsent(
                room, k -> Collections.newSetFromMap(new ConcurrentHashMap<>()));
        Runnable tracked = new Runnable() {
            @Override public void run() {
                unsub.run();
                Set<Runnable> s = roomUnsubs.get(room);
                if (s != null) s.remove(this);
            }
        };
        set.add(tracked);
        return tracked;
    }

    /**
     * Wrap a callback so payloads that carry an ID field ({@code idField} or
     * the optional {@code altIdField}) not matching {@code expected} are dropped.
     * Payloads without the field pass through unchanged, so multi-room
     * subscribers never receive cross-room events.
     */
    private static Consumer<JsonNode> filterById(
            String idField, String altIdField, String expected, Consumer<JsonNode> cb) {
        return payload -> {
            if (payload != null && payload.isObject()) {
                JsonNode id = payload.get(idField);
                if ((id == null || id.isNull()) && altIdField != null) id = payload.get(altIdField);
                if (id != null && !id.isNull() && !expected.equals(id.asText())) return;
            }
            cb.accept(payload);
        };
    }

    private void emit(String event, JsonNode payload) {
        Set<Consumer<JsonNode>> set = listeners.get(event);
        if (set == null) return;
        for (Consumer<JsonNode> l : set) {
            try { l.accept(payload); } catch (Exception ignored) {}
        }
    }

    /** Decode a Socket.IO event array and dispatch. */
    private void handleSioEvent(String sioRaw) {
        // sioRaw starts with "2" followed by JSON array: ["eventName", payload]
        try {
            String json = sioRaw.substring(1); // strip SIO type digit
            JsonNode arr = MAPPER.readTree(json);
            if (!arr.isArray() || arr.size() < 2) return;
            String eventName = arr.get(0).asText();
            JsonNode payload = arr.get(1);
            emit(eventName, payload);
        } catch (Exception ignored) {}
    }

    /** Handle a single Engine.IO packet string. */
    private void handlePacket(String raw) {
        if (raw == null || raw.isEmpty()) return;
        char type = raw.charAt(0);
        if (type == EIO_PING) {
            wsSend(String.valueOf(EIO_PONG));
            return;
        }
        if (type != EIO_MESSAGE) return;
        String sioRaw = raw.substring(1);
        if (sioRaw.isEmpty()) return;
        int sioType;
        try { sioType = Integer.parseInt(String.valueOf(sioRaw.charAt(0))); }
        catch (NumberFormatException e) { return; }

        if (sioType == SIO_EVENT && sioRaw.length() > 1) {
            handleSioEvent(sioRaw);
        }
        // SIO_CONNECT (0) is handled implicitly by the connect flow
    }

    /** Parse Engine.IO concatenated poll response ({@code "<len>:<packet>..."} format). */
    private void parsePollResponse(String text) {
        int i = 0;
        while (i < text.length()) {
            int colon = text.indexOf(':', i);
            if (colon == -1) { handlePacket(text.substring(i)); break; }
            int len;
            try { len = Integer.parseInt(text, i, colon, 10); }
            catch (NumberFormatException e) { handlePacket(text.substring(i)); break; }
            String packet = text.substring(colon + 1, Math.min(colon + 1 + len, text.length()));
            handlePacket(packet);
            i = colon + 1 + len;
        }
    }

    /**
     * Emit the correct server-side join event for a room string.
     * <ul>
     *   <li>{@code run:<id>}     → emit {@code joinRunRoom} with bare {@code <id>}</li>
     *   <li>{@code project:<p>}  → emit {@code joinProjectRoom} with bare {@code <p>}</li>
     *   <li>anything else        → emit {@code joinRunRoom} with the full string (safe fallback)</li>
     * </ul>
     */
    private void emitJoin(String room) {
        if (room.startsWith("run:")) {
            sioEmit("joinRunRoom", room.substring(4));
        } else if (room.startsWith("project:")) {
            sioEmit("joinProjectRoom", room.substring(8));
        } else {
            sioEmit("joinRunRoom", room); // safe fallback
        }
    }

    private void joinRoom(String room) {
        ensureConnect();
        if (joinedRooms.add(room) && sid != null) {
            emitJoin(room);
        }
    }

    private void leaveRoom(String room) {
        // Remove from the rejoin-on-reconnect list …
        joinedRooms.remove(room);
        // … drop every listener registered for this room …
        Set<Runnable> unsubs = roomUnsubs.remove(room);
        if (unsubs != null) {
            for (Runnable r : unsubs) {
                try { r.run(); } catch (Exception ignored) {}
            }
        }
        // … and tell the server. Server expects leaveRoom with the FULL room
        // string ("run:<id>" or "project:<p>").
        if (sid != null) sioEmit("leaveRoom", room);
    }

    private void replayJoins() {
        for (String room : joinedRooms) {
            emitJoin(room);
        }
    }

    private void wsSend(String data) {
        WebSocketClient w = ws;
        if (w != null && w.isOpen()) {
            try { w.send(data); } catch (Exception ignored) {}
        }
    }

    /** Build a Socket.IO event packet string. */
    private void sioEmit(String event, Object... args) {
        try {
            List<Object> list = new ArrayList<>();
            list.add(event);
            Collections.addAll(list, args);
            String json = MAPPER.writeValueAsString(list);
            wsSend(EIO_MESSAGE + "" + SIO_EVENT + json);
        } catch (Exception ignored) {}
    }

    /** Build the Socket.IO connect packet with Bearer auth. */
    private String sioConnectPacket() {
        try {
            String authJson = MAPPER.writeValueAsString(
                    Collections.singletonMap("token", apiKey));
            return EIO_MESSAGE + "" + SIO_CONNECT + authJson;
        } catch (Exception e) {
            return EIO_MESSAGE + "" + SIO_CONNECT + "{\"auth\":{}}";
        }
    }

    private String pollUrl() {
        return baseUrl + "/api/v1/events/?EIO=4&transport=polling"
                + (sid != null ? "&sid=" + URLEncoder.encode(sid, StandardCharsets.UTF_8) : "");
    }

    private String wsUrl() {
        String wsBase = baseUrl.replace("https://", "wss://").replace("http://", "ws://");
        return wsBase + "/api/v1/events/?EIO=4&transport=websocket"
                + (sid != null ? "&sid=" + URLEncoder.encode(sid, StandardCharsets.UTF_8) : "");
    }

    /** Kick off the lazy connect exactly once, on the first subscription/join. */
    private void ensureConnect() {
        if (!closed && connectRequested.compareAndSet(false, true)) {
            scheduleConnect(0);
        }
    }

    private void scheduleConnect(long delayMs) {
        if (closed) return;
        scheduler.schedule(this::connect, delayMs, TimeUnit.MILLISECONDS);
    }

    private void connect() {
        if (closed) return;
        try {
            // Step 1: Engine.IO HTTP handshake
            HttpRequest handshake = HttpRequest.newBuilder()
                    .uri(URI.create(pollUrl()))
                    .header("Authorization", "Bearer " + apiKey)
                    .GET()
                    .build();
            HttpResponse<String> resp = httpClient.send(handshake, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            String text = resp.body();
            int jsonStart = text.indexOf('{');
            if (jsonStart == -1) throw new IllegalStateException("Unexpected EIO handshake: " + text);
            JsonNode openData = MAPPER.readTree(text.substring(jsonStart));
            sid = openData.get("sid").asText();

            // Step 2: Socket.IO connect packet via POST
            HttpRequest connectPost = HttpRequest.newBuilder()
                    .uri(URI.create(pollUrl()))
                    .header("Authorization", "Bearer " + apiKey)
                    .header("Content-Type", "text/plain;charset=UTF-8")
                    .POST(HttpRequest.BodyPublishers.ofString(sioConnectPacket(), StandardCharsets.UTF_8))
                    .build();
            httpClient.send(connectPost, HttpResponse.BodyHandlers.discarding());

            // Step 3: Upgrade to WebSocket
            upgradeToWs();

            // Replay queued room joins
            replayJoins();

        } catch (Exception e) {
            sid = null;
            if (!closed) {
                LOG.log(Level.WARNING, "Melaya events: connect failed ("
                        + e.getClass().getSimpleName() + ": " + e.getMessage()
                        + "); retrying in " + CONNECT_RETRY_MS + " ms");
                scheduleConnect(CONNECT_RETRY_MS);
            }
        }
    }

    private void upgradeToWs() {
        if (closed) return;
        String url = wsUrl();
        WebSocketClient client;
        boolean connected;
        try {
            client = new WebSocketClient(new URI(url)) {
                @Override public void onOpen(ServerHandshake h) {
                    // Engine.IO WebSocket upgrade probe
                    send("2probe");
                    send("5"); // upgrade packet
                }
                @Override public void onMessage(String message) {
                    handlePacket(message);
                }
                @Override public void onClose(int code, String reason, boolean remote) {
                    ws = null;
                    if (!closed) {
                        sid = null;
                        long delay = RECONNECT_BACKOFF_MS[
                                Math.min(reconnectAttempt, RECONNECT_BACKOFF_MS.length - 1)];
                        reconnectAttempt++;
                        LOG.log(Level.WARNING, "Melaya events: WebSocket closed (code=" + code
                                + ", remote=" + remote + "); reconnecting in " + delay + " ms");
                        scheduleConnect(delay);
                    }
                }
                @Override public void onError(Exception ex) {
                    // onClose follows — reconnect is scheduled there
                    LOG.log(Level.WARNING, "Melaya events: WebSocket error ("
                            + ex.getClass().getSimpleName() + ": " + ex.getMessage() + ")");
                }
            };
            // Daemon: the WebSocket read/write threads must never keep the JVM alive.
            client.setDaemon(true);
            connected = client.connectBlocking(10, TimeUnit.SECONDS);
        } catch (Exception e) {
            if (!closed) {
                long delay = RECONNECT_BACKOFF_MS[
                        Math.min(reconnectAttempt, RECONNECT_BACKOFF_MS.length - 1)];
                reconnectAttempt++;
                LOG.log(Level.WARNING, "Melaya events: WebSocket upgrade failed ("
                        + e.getClass().getSimpleName() + ": " + e.getMessage()
                        + "); reconnecting in " + delay + " ms");
                scheduleConnect(delay);
            }
            return;
        }
        if (!connected) {
            // connectBlocking timed out or the handshake was rejected
            if (!closed) {
                long delay = RECONNECT_BACKOFF_MS[
                        Math.min(reconnectAttempt, RECONNECT_BACKOFF_MS.length - 1)];
                reconnectAttempt++;
                LOG.log(Level.WARNING,
                        "Melaya events: WebSocket did not open within 10 s; reconnecting in " + delay + " ms");
                scheduleConnect(delay);
            }
            return;
        }
        reconnectAttempt = 0; // successful upgrade — reset backoff
        ws = client;
    }
}

File: packages/sdk-kotlin/src/main/kotlin/org/melaya/MelayaEvents.kt
package org.melaya

import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONArray
import org.json.JSONObject
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicReference

/**
 * Platform real-time events via Socket.IO v4 at `/api/v1/events`.
 *
 * Implements the Engine.IO v4 + Socket.IO v4 wire protocol using OkHttp's
 * WebSocket client. The connection strategy mirrors the TypeScript SDK:
 *
 *   1. Perform the Engine.IO HTTP long-poll handshake to obtain a `sid`.
 *   2. Post the Socket.IO connect packet with the Bearer token as auth data.
 *   3. Upgrade to WebSocket.
 *   4. Replay queued room joins (and reset the back-off counter).
 *   5. Reconnect with exponential back-off on failure or disconnect,
 *      logging each failure via `java.util.logging`.
 *
 * The connection is opened **lazily** on the first subscription or room join —
 * constructing [MelayaEvents] (or [Melaya]) performs no events I/O, so
 * REST-only usage never opens a socket. All internal threads are daemon
 * threads, so an open events connection does not prevent JVM exit; still call
 * [close] for a prompt, clean shutdown.
 *
 * Room semantics (server-side):
 *   - `run:<runId>`        — events for a specific pipeline run
 *   - `project:<project>`  — all events in a project (runs + pipeline CRUD)
 *   - `hitl:user:<userId>` — HITL approval events; joined automatically by the server
 *
 * @example
 * ```kotlin
 * val events = melaya.events
 * val unsubRun     = events.onRunUpdate("run-123") { e -> println(e.optString("event_type")) }
 * val unsubHitl    = events.onHitlApproval { e -> println("HITL: ${e.optString("type")}") }
 * val unsubProject = events.onProjectEvent("my-project") { e -> println(e) }
 *
 * // Clean up
 * unsubRun()
 * events.leaveRun("run-123")
 * events.close()
 * ```
 */
class MelayaEvents internal constructor(
    private val baseUrl: String,
    /** The `mk_` platform API key — sent as Bearer token in the connect packet. */
    private val apiKey: String,
    private val okHttp: OkHttpClient,
) : AutoCloseable {

    // ── Engine.IO / Socket.IO packet type constants ──────────────────────────

    private val EIO_OPEN    = '0'
    private val EIO_PING    = '2'
    private val EIO_PONG    = '3'
    private val EIO_MESSAGE = '4'

    private val SIO_CONNECT = 0
    private val SIO_EVENT   = 2

    // ── State ────────────────────────────────────────────────────────────────

    private val log = java.util.logging.Logger.getLogger(MelayaEvents::class.java.name)

    private val closed        = AtomicBoolean(false)
    private val started       = AtomicBoolean(false)
    private val sid           = AtomicReference<String?>(null)
    private val ws            = AtomicReference<WebSocket?>(null)
    private var reconnectAttempt = 0

    /** rooms currently joined (or pending join after connect) — replayed on reconnect */
    private val joinedRooms = ConcurrentHashMap.newKeySet<String>()

    /** event name → list of callbacks */
    private val listeners = ConcurrentHashMap<String, CopyOnWriteArrayList<(JSONObject) -> Unit>>()

    /** room → (event, callback) pairs registered through that room's on* helpers */
    private val roomSubs = ConcurrentHashMap<String, CopyOnWriteArrayList<Pair<String, (JSONObject) -> Unit>>>()

    // ── Public subscription API ───────────────────────────────────────────────

    /**
     * Subscribe to run events for a specific pipeline run.
     * Automatically joins the `run:<runId>` Socket.IO room.
     *
     * When the payload carries a `runId` it is filtered against [runId], so
     * subscribers to multiple run rooms never receive cross-room events.
     *
     * Returns an unsubscribe function.
     */
    fun onRunUpdate(runId: String, cb: (JSONObject) -> Unit): () -> Unit {
        joinRoom("run:$runId")
        return on("pushEvent", runFiltered(runId, cb), room = "run:$runId")
    }

    /**
     * Subscribe to init-phase progress events for a run.
     * Automatically joins the `run:<runId>` room.
     *
     * When the payload carries a `runId` it is filtered against [runId].
     *
     * Returns an unsubscribe function.
     */
    fun onInitPhase(runId: String, cb: (JSONObject) -> Unit): () -> Unit {
        joinRoom("run:$runId")
        return on("pushInitPhase", runFiltered(runId, cb), room = "run:$runId")
    }

    /**
     * Subscribe to all events in a project (all runs, pipeline CRUD).
     * Automatically joins the `project:<project>` room.
     *
     * When the payload carries a `projectId` it is filtered against [project],
     * so subscribers to multiple project rooms never receive cross-room events.
     *
     * Returns an unsubscribe function.
     */
    fun onProjectEvent(project: String, cb: (JSONObject) -> Unit): () -> Unit {
        joinRoom("project:$project")
        return on("pushEvent", projectFiltered(project, cb), room = "project:$project")
    }

    /**
     * Subscribe to HITL approval invalidation events.
     * The server automatically joins authenticated sockets to the
     * `hitl:user:<userId>` room on connect — no explicit join needed.
     *
     * Returns an unsubscribe function.
     */
    fun onHitlApproval(cb: (JSONObject) -> Unit): () -> Unit {
        return on("pushHitlApprovals", cb)
    }

    /** Subscribe to pipeline creation events within a project room. */
    fun onPipelineCreated(project: String, cb: (JSONObject) -> Unit): () -> Unit {
        joinRoom("project:$project")
        return on("pipelineCreated", projectFiltered(project, cb), room = "project:$project")
    }

    /** Subscribe to pipeline update events within a project room. */
    fun onPipelineUpdated(project: String, cb: (JSONObject) -> Unit): () -> Unit {
        joinRoom("project:$project")
        return on("pipelineUpdated", projectFiltered(project, cb), room = "project:$project")
    }

    /** Subscribe to pipeline deletion events within a project room. */
    fun onPipelineDeleted(project: String, cb: (JSONObject) -> Unit): () -> Unit {
        joinRoom("project:$project")
        return on("pipelineDeleted", projectFiltered(project, cb), room = "project:$project")
    }

    /**
     * Stop listening to a specific run: leaves the `run:<runId>` room, removes
     * every listener registered through that room's `on*` helpers, and drops
     * the room from the rejoin-on-reconnect list.
     */
    fun leaveRun(runId: String) = leaveRoom("run:$runId")

    /**
     * Leave a project room: removes every listener registered through that
     * room's `on*` helpers and drops the room from the rejoin-on-reconnect list.
     */
    fun leaveProject(project: String) = leaveRoom("project:$project")

    /** Close the Socket.IO connection and release all listeners. */
    override fun close() {
        closed.set(true)
        ws.getAndSet(null)?.close(1000, "client closed")
        listeners.clear()
        roomSubs.clear()
        joinedRooms.clear()
    }

    // ── Internal helpers ──────────────────────────────────────────────────────

    /**
     * Lazily open the Socket.IO connection on the first subscription or room
     * join. Constructing [MelayaEvents] performs no network I/O, so REST-only
     * usage never opens (or waits on) an events socket.
     */
    private fun ensureStarted() {
        if (closed.get()) return
        if (started.compareAndSet(false, true)) connect()
    }

    /** Wrap [cb] so it only fires when the payload's `runId` (if present) matches [runId]. */
    private fun runFiltered(runId: String, cb: (JSONObject) -> Unit): (JSONObject) -> Unit = { e ->
        val id = e.optString("runId", "")
        if (id.isEmpty() || id == runId) cb(e)
    }

    /** Wrap [cb] so it only fires when the payload's `projectId` (if present) matches [project]. */
    private fun projectFiltered(project: String, cb: (JSONObject) -> Unit): (JSONObject) -> Unit = { e ->
        val id = e.optString("projectId", "")
        if (id.isEmpty() || id == project) cb(e)
    }

    private fun on(event: String, cb: (JSONObject) -> Unit, room: String? = null): () -> Unit {
        ensureStarted()
        listeners.getOrPut(event) { CopyOnWriteArrayList() }.add(cb)
        if (room != null) roomSubs.getOrPut(room) { CopyOnWriteArrayList() }.add(event to cb)
        return {
            listeners[event]?.remove(cb)
            if (room != null) roomSubs[room]?.removeAll { it.second === cb }
        }
    }

    private fun emit(event: String, payload: JSONObject) {
        listeners[event]?.forEach { cb ->
            try { cb(payload) } catch (_: Exception) { /* listener errors must not crash the client */ }
        }
    }

    private fun handleSioEvent(args: JSONArray) {
        if (args.length() < 2) return
        val event   = args.optString(0) ?: return
        val payload = args.optJSONObject(1) ?: JSONObject()
        emit(event, payload)
    }

    private fun handlePacket(raw: String) {
        if (raw.isEmpty()) return
        when (raw[0]) {
            EIO_PING -> wsSend(EIO_PONG.toString())
            EIO_MESSAGE -> {
                val sioRaw = raw.drop(1)
                if (sioRaw.isEmpty()) return
                val sioType = sioRaw[0].digitToIntOrNull() ?: return
                if (sioType != SIO_EVENT) return
                try {
                    val args = JSONArray(sioRaw.drop(1))
                    handleSioEvent(args)
                } catch (_: Exception) { /* ignore malformed packets */ }
            }
            else -> { /* ignore */ }
        }
    }

    /**
     * Map a logical room string to the correct server event name + bare arg.
     *
     * Server contract (server/src/index.ts):
     *   "run:<runId>"       -> emit joinRunRoom(runId)
     *   "project:<project>" -> emit joinProjectRoom(project)
     *   anything else       -> no-op (hitl room is auto-joined by the server)
     */
    private fun emitJoin(room: String) {
        when {
            room.startsWith("run:")     -> sioEmit("joinRunRoom",     room.substring(4))
            room.startsWith("project:") -> sioEmit("joinProjectRoom", room.substring(8))
            // hitl:user:<userId> is auto-joined by the server on connect — no emit needed
        }
    }

    private fun joinRoom(room: String) {
        ensureStarted()
        if (joinedRooms.add(room) && sid.get() != null) {
            emitJoin(room)
        }
        // if not connected yet, joinedRooms will be replayed in replayJoins()
    }

    private fun leaveRoom(room: String) {
        joinedRooms.remove(room)  // no longer rejoined on reconnect
        // Drop every listener registered through this room's on* helpers.
        roomSubs.remove(room)?.forEach { (event, cb) -> listeners[event]?.remove(cb) }
        // Server expects leaveRoom with the FULL room string ("run:<runId>" / "project:<project>")
        if (sid.get() != null) sioEmit("leaveRoom", room)
    }

    private fun replayJoins() {
        joinedRooms.forEach { room -> emitJoin(room) }
    }

    private fun sioEmit(event: String, vararg args: Any) {
        val arr = JSONArray().put(event)
        args.forEach { arr.put(it) }
        wsSend("$EIO_MESSAGE$SIO_EVENT$arr")
    }

    private fun sioConnectPacket(): String {
        val authData = JSONObject().put("token", apiKey)
        return "$EIO_MESSAGE${SIO_CONNECT}$authData"
    }

    private fun wsSend(data: String) {
        ws.get()?.send(data)
    }

    // ── Connection lifecycle ──────────────────────────────────────────────────

    private fun pollUrl(extra: Map<String, String> = emptyMap()): String {
        val base = baseUrl.trimEnd('/')
        val sb = StringBuilder("$base/api/v1/events/?EIO=4&transport=polling")
        sid.get()?.let { sb.append("&sid=${enc(it)}") }
        extra.forEach { (k, v) -> sb.append("&${enc(k)}=${enc(v)}") }
        return sb.toString()
    }

    private fun wsUrl(): String {
        val base = baseUrl.trimEnd('/')
            .replace("^https://".toRegex(), "wss://")
            .replace("^http://".toRegex(),  "ws://")
        val sb = StringBuilder("$base/api/v1/events/?EIO=4&transport=websocket")
        sid.get()?.let { sb.append("&sid=${enc(it)}") }
        return sb.toString()
    }

    /** Bounded exponential backoff: 1s, 2s, 4s, 8s, … capped at 30s, with ±20% jitter. */
    private fun reconnectDelayMs(): Long {
        val base = minOf(1_000L shl minOf(reconnectAttempt, 5), 30_000L)
        val jitter = (base * 0.2 * Math.random()).toLong()
        reconnectAttempt++
        return base + jitter
    }

    /**
     * Log the failure and retry after a backoff delay on a daemon thread —
     * never blocks the calling (OkHttp) thread and never keeps the JVM alive.
     */
    private fun scheduleReconnect(reason: String) {
        if (closed.get()) return
        val delayMs = reconnectDelayMs()
        log.warning("Melaya events connection lost ($reason); reconnecting in ${delayMs}ms (attempt $reconnectAttempt)")
        Thread {
            try { Thread.sleep(delayMs) } catch (_: InterruptedException) { return@Thread }
            connect()
        }.also { it.isDaemon = true; it.name = "melaya-events-reconnect" }.start()
    }

    private fun connect() {
        if (closed.get()) return
        Thread {
            try {
                // Step 1: Engine.IO long-poll handshake
                val handshakeReq = Request.Builder()
                    .url(pollUrl())
                    .header("Authorization", "Bearer $apiKey")
                    .get()
                    .build()
                val handshakeResp: Response = okHttp.newCall(handshakeReq).execute()
                val text = handshakeResp.body?.string() ?: throw IllegalStateException("Empty EIO handshake")
                val jsonStart = text.indexOf('{')
                if (jsonStart == -1) throw IllegalStateException("Unexpected EIO response: $text")
                val openData = JSONObject(text.substring(jsonStart))
                sid.set(openData.getString("sid"))

                // Step 2: Send SIO connect packet with auth
                val connectBody = sioConnectPacket()
                    .toRequestBody("text/plain;charset=UTF-8".toMediaType())
                val connectReq = Request.Builder()
                    .url(pollUrl())
                    .header("Authorization", "Bearer $apiKey")
                    .post(connectBody)
                    .build()
                okHttp.newCall(connectReq).execute().close()

                // Step 3: Upgrade to WebSocket
                if (!closed.get()) upgradeToWs()

                // Step 4: Replay pending room joins; a successful connection resets the backoff
                reconnectAttempt = 0
                replayJoins()

            } catch (e: Exception) {
                scheduleReconnect("${e.javaClass.simpleName}: ${e.message ?: "handshake failed"}")
            }
        }.also { it.isDaemon = true; it.name = "melaya-events-connect" }.start()
    }

    private fun upgradeToWs() {
        val req = Request.Builder()
            .url(wsUrl())
            .header("Authorization", "Bearer $apiKey")
            .build()

        val listener = object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                // Engine.IO WebSocket upgrade probe
                webSocket.send("2probe")
                webSocket.send("5")
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                handlePacket(text)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                ws.compareAndSet(webSocket, null)
                if (!closed.get()) {
                    sid.set(null)
                    scheduleReconnect("websocket failure: ${t.javaClass.simpleName}")
                }
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                ws.compareAndSet(webSocket, null)
                if (!closed.get()) {
                    sid.set(null)
                    scheduleReconnect("websocket closed (code=$code)")
                }
            }
        }

        ws.set(okHttp.newWebSocket(req, listener))
    }

    private fun enc(s: String) = java.net.URLEncoder.encode(s, "UTF-8")

}
